### PR TITLE
IC-1459: Ability to add initial assessment to referral

### PR DESCRIPTION
**Pull request by rgforsyth**
## What does this pull request do?

Adds tables for supplier assessments.
Automatically supplier assessment with no appointments when a referral is sent.
Adds endpoints to PATCH(add/update) a supplier assessment appointment & GET supplier assessment appointment.

## What is the intent behind these changes?

Support supplier assessments

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/ActionPlanSessionsDTO.kt [hunk 1]
```diff
@@ -1,13 +1,14 @@
...
+import com.fasterxml.jackson.annotation.JsonProperty
 import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.ActionPlanSession
 import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.Attended
 import java.time.OffsetDateTime
 import java.util.UUID
 
 data class UpdateAppointmentDTO(
   val appointmentTime: OffsetDateTime,
-  val durationInMinutes: Int,
+  @JsonProperty(required = true) val durationInMinutes: Int,
```
Comment (rgforsyth) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/ActionPlanSessionsDTO.kt:11]: This is needed, otherwise the duration will default to '0' if it isn't included. This will then set the duration of the appointment to 0.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/Appointment.kt [hunk 2]
```diff
@@ -46,6 +46,6 @@ enum class Attended {
 
   @JsonValue
   open fun toLower(): String? {
-    return this.toString().lowercase()
+    return this.toString().toLowerCase()
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/Appointment.kt:49]: this method is deprecated since kotlin 1.5 - the existing method is the newer version

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/SupplierAssessment.kt [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+@Entity
+@Table(name = "supplier_assessment")
+data class SupplierAssessment(
+  @Id val id: UUID,
+  @NotNull @OneToOne val referral: Referral,
+
+  @NotNull @OneToMany @Fetch(FetchMode.JOIN) val appointments: MutableSet<Appointment> = mutableSetOf(),
+) {
+  val appointment: Appointment
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/SupplierAssessment.kt:21]: please can we name this simliarly to the action plan one `currentAppointment`? one of the intentions with that was to make a future find & replace easier when we actually have the possibilty there are multiple appointments per thing.

File: src/main/resources/db/migration/V1_57__supplier_assessment.sql [hunk 4]
```diff
@@ -0,0 +1,15 @@
+create table supplier_assessment(
+    id uuid not null,
+    referral_id uuid,
```
Comment (tomsmyers) [inline src/main/resources/db/migration/V1_57__supplier_assessment.sql]: can we add a unique key on referral id - meaning there can only ever be one supplier assessment per referral?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt [hunk 5]
```diff
@@ -159,6 +160,7 @@ class ReferralService(
 
     val sentReferral = referralRepository.save(referral)
     eventPublisher.referralSentEvent(sentReferral)
+    supplierAssessmentService.createInitialAssessment(referral, user)
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt]: naming issue here - 'initial assessment' isn't in use, i think it's just 'supplier assessment' ?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/SupplierAssessmentService.kt [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+@Service
+@Transactional
+class SupplierAssessmentService(
+  val supplierAssessmentRepository: SupplierAssessmentRepository,
+  val referralRepository: ReferralRepository,
+) {
+  fun createInitialAssessment(
+    referral: Referral,
+    createdByUser: AuthUser
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/SupplierAssessmentService.kt]: code smell here with this not being used ...

since this is created automatically does it make sense to have a createdBy field ?

we have a createdBy field on the appointments, so perhaps that's enough?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/SupplierAssessmentController.kt [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.service.SupplierAssessmentService
+import java.util.UUID
+
+@RestController
+class SupplierAssessmentController(
+  private val supplierAssessmentService: SupplierAssessmentService,
+  private val referralService: ReferralService,
+  private val userMapper: UserMapper,
+) {
+  @PatchMapping("/sent-referral/{id}/supplier-assessment-appointment")
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/SupplierAssessmentController.kt:25]: this seems an odd URL to have outside of the referrals controller?

how come this isn't updating the supplier assessment by id?

i would expect this to be more like `POST /supplier-assessment/{id}/appointment`

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/ActionPlanSessionsDTO.kt (hunk 1 above)
Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/ActionPlanSessionsDTO.kt:11]: is there anywhere else we have not nullable basic types?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/SupplierAssessmentDTO.kt [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+  val createdAt: OffsetDateTime,
+  val createdBy: AuthUserDTO,
+) {
+  companion object {
+    fun from(supplierAssessment: SupplierAssessment): SupplierAssessmentDTO {
+      return SupplierAssessmentDTO(
+        id = supplierAssessment.id,
+        appointmentTime = supplierAssessment.appointment.appointmentTime,
+        durationInMinutes = supplierAssessment.appointment.durationInMinutes,
+        createdAt = supplierAssessment.appointment.createdAt,
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/SupplierAssessmentDTO.kt:21]: see my comment about createdAt/By elsewhere. let's not follow the same pattern here just to match the other. 

why don't we use this opportunity to split the appointment in the DTO?

Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/SupplierAssessmentDTO.kt:21]: i.e. 

{
  id: ...
  appointment: {
    durationInMinutes: ...,
    appointmentTime: ...,
    createdBy: ...,
    deliusAppointmentId: ...,
    ...
    }
 }

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/SupplierAssessmentService.kt [hunk 9]
```diff
@@ -27,4 +32,29 @@ class SupplierAssessmentService(
...
     return referralRepository.save(referral)
   }
+
+  fun updateSupplierAssessmentAppointment(
+    referral: Referral,
+    durationInMinutes: Int,
+    appointmentTime: OffsetDateTime,
+    createdByUser: AuthUser
+  ): SupplierAssessment {
+    if (referral.supplierAssessment!!.appointments.size == 0) {
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/SupplierAssessmentService.kt:42]: would be good to have more explicit error handling around referral.supplierAssessment

either delegate this to the caller by taking a not null supplierAssessment in the signature, or throw an error with an explicit message like "no supplier assessment for rererral id ${referral.id}"

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/SupplierAssessmentService.kt [hunk 10]
```diff
@@ -27,4 +32,29 @@ class SupplierAssessmentService(
...
+            id = UUID.randomUUID(),
+            appointmentTime = appointmentTime,
+            durationInMinutes = durationInMinutes,
+            createdBy = authUserRepository.save(createdByUser),
+            createdAt = OffsetDateTime.now()
+          )
+        )
+      )
+    } else {
+      referral.supplierAssessment!!.appointment.durationInMinutes = durationInMinutes
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/SupplierAssessmentService.kt:55]: do we want the same endpoint for both scheduling and rescheduling appointments?

could we delegate this out to an appointments service that both this and the action plan session code can call?

e.g. `AppointmentsService.bookOrRescheduleAppointment(appointment: Appointment?, durationInMinutes: Int, appointmentTime: OffsetDateTime, actor: AuthUser)`

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/SupplierAssessmentController.kt [hunk 11]
```diff
@@ -39,4 +41,21 @@ class SupplierAssessmentController(
       )
     )
   }
+
+  @GetMapping("sent-referral/{id}/supplier-assessment-appointment")
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/SupplierAssessmentController.kt:45]: same comment on the naming as before. 

if the supplier assessment id is on the referral dto then could we have `GET /supplier-assessment/{id}`

**Review by tomsmyers — COMMENTED** [review]
nice work! a few comments, some more serious than others e.g. URLs

we also need to add in stuff for delius bookings which i didn't see in here

**Comment by tomsmyers** [comment]
@sldblog what are your thoughts on the URLs for supplier assessments in the context of this PR?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/SupplierAssessmentController.kt (hunk 7 above)
Comment (sldblog) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/SupplierAssessmentController.kt:25]: thinking out loud, about suppler assessment appointments, we need to

- book them
- reschedule them (re-book)
- record attendance, behaviour and feedback
- and list them

all of these are in the context of the referral

Since we already have `GET /action-plan?referralId=...` pattern, I think it's ok to have

- `GET /supplier-assessment?referralId=...`
- `POST /supplier-assessment` with `referralId`
- `PATCH /supplier-assessment/{id}`

It would be really good if we could mention the key domain verbs here somehow -- `book`, `attendance`, etc.

Comment (rgforsyth) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/SupplierAssessmentController.kt:25]: the endpoints that were going to be created (according to contract) were:

get saa
patch saa
record saa attendance
record saa behaviour
submit saa session feedback.

So, following the above, would you like these to be:
GET: /supplier-assessment-appointment/{id}
PATCH: /supplier-assessment-appointment/{id}
POST: /supplier-assessment-appointment/{id}/record-attendance
POST: /supplier-assessment-appointment/{id}/record-behaviour
POST: /supplier-assessment-appointment/{id}/submit

Comment (rgforsyth) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/SupplierAssessmentController.kt:25]: Decision from meeting:

supplier assessment URLs
# ReferralsController
GET /referrals/{referralId}/supplier-assessment -> {id: …, appointments: [], currentAppointmentId: … } (SupplierAssessmentDTO) = get supplier assessment for referral
# SupplierAssessmentsController
PUT /supplier-assessment/{supplierAssessmentId}/schedule-appointment -> AppointmentDTO = book or reschedule appointment
	- if there is no appointment, book a new one.
	- if there is an appointment with no attendance data, reschedule
	- if there is an appointment with attendance but didn’t attend, book new one
	- if there is an appointment with attendance but did attend, error
# AppointmentsController
PUT /appointments/{id}/record-attendance - > AppointmentDTO
PUT /appointments/{id}/record-behaviour - > AppointmentDTO
POST /appointments/{id}/submit-feedback -> AppointmentDTO

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/SupplierAssessmentService.kt [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+
+  fun scheduleOrUpdateSupplierAssessmentAppointment(
+    supplierAssessmentId: UUID,
+    durationInMinutes: Int,
+    appointmentTime: OffsetDateTime,
+    createdByUser: AuthUser
+  ): Appointment {
+    val supplierAssessment = getSupplierAssessmentById(supplierAssessmentId)
+
+    if (supplierAssessment.appointments.size == 0) {
```
Comment (mahalrmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/SupplierAssessmentService.kt:45]: Why is this needed? Isn't the appointment always in existence once referral is sent?

Comment (rgforsyth) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/SupplierAssessmentService.kt:45]: The supplier assessment gets created. The appointment is not actually created until the schedule endpoint is called for the first time.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/SupplierAssessmentDTO.kt [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+        id = supplierAssessment.id,
+        appointments = AppointmentDTO.from(supplierAssessment.appointments),
+        currentAppointmentId = supplierAssessment.currentAppointment.id,
+        referralId = supplierAssessment.referral.id
+      )
+    }
+  }
+}
+
+data class SupplierAssessmentAppointmentDTO(
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/SupplierAssessmentDTO.kt:26]: is this used anywhere?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/SupplierAssessmentDTO.kt [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.Appointment
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.SupplierAssessment
+import java.time.OffsetDateTime
+import java.util.UUID
+
+data class SupplierAssessmentDTO(
+  val id: UUID,
+  val appointments: List<AppointmentDTO>,
+  val currentAppointmentId: UUID,
+  val referralId: UUID,
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/SupplierAssessmentDTO.kt:12]: is there a use-case for including this field?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/ReferralController.kt [hunk 15]
```diff
@@ -182,6 +183,23 @@ class ReferralController(
...
+  fun getSupplierAssessmentAppointment(
+    @PathVariable id: UUID,
+    authentication: JwtAuthenticationToken,
+  ): SupplierAssessmentDTO {
+    val user = userMapper.fromToken(authentication)
+
+    val sentReferral = referralService.getSentReferralForUser(id, user)
+      ?: throw ResponseStatusException(HttpStatus.NOT_FOUND, "sent referral not found [id=$id]")
+
+    if (sentReferral.supplierAssessment!!.appointments.size == 0) {
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/ReferralController.kt:196]: what's the reason this is here?

Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/ReferralController.kt:196]: also i don't like the not null assertion on supplierAssessment. we can get a not null reference to the SA like this:

```
val assessment = sentReferral.supplierAssessment ?: throw(not found....)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/SupplierAssessmentService.kt [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+    supplierAssessment: SupplierAssessment,
+    durationInMinutes: Int,
+    appointmentTime: OffsetDateTime,
+    createdByUser: AuthUser
+  ): Appointment{
+    return engagementService.createOrUpdateEngagement(supplierAssessment, durationInMinutes, appointmentTime, createdByUser)
+  }
+
+  fun getSupplierAssessmentById(supplierAssessmentId: UUID): SupplierAssessment {
+    return supplierAssessmentRepository.findById(supplierAssessmentId).orElseThrow {
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/SupplierAssessmentService.kt:46]: nit: usage of java optional discouraged in kotlin. use `findByIdOrNull` instead

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/ActionPlanSessionsDTO.kt (hunk 1 above)
Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/ActionPlanSessionsDTO.kt:12]: @lawrence-forooghian can we have a contract test to validate this behaviour?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/AppointmentDTO.kt [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+      return AppointmentDTO(
+        id = appointment.id,
+        appointmentTime = appointment.appointmentTime,
+        durationInMinutes = appointment.durationInMinutes,
+        sessionFeedback = SessionFeedbackDTO.from(
+          appointment.attended,
+          appointment.additionalAttendanceInformation,
+          appointment.attendanceBehaviour,
+          appointment.notifyPPOfAttendanceBehaviour,
+          appointment.attendanceSubmittedAt != null,
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/AppointmentDTO.kt:24]: ⛔️ blocker: this should be `appointment.appointmentFeedbackSubmittedAt`

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/AppointmentService.kt [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    val appointmentWithNonAttendance = appointment.attended == Attended.NO
+    if (appointmentWithNonAttendance) {
+      return createAppointment(durationInMinutes, appointmentTime, createdByUser)
+    }
+    throw IllegalStateException("Is it not possible to update an appointment that has already been attended")
+  }
+
+  private fun createAppointment(
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/AppointmentService.kt:42]: small thing - i would expect this to be the place where the delius bookings are updated. is that coming later?

**Review by tomsmyers — APPROVED** [review]
great work - one major thing that needs fixing before merge (session feedback submitted flag).

would also be nice to have a unit test for `currentAppointment` after create/update.

✅

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/AppointmentService.kt (hunk 18 above)
Comment (rgforsyth) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/AppointmentService.kt:42]: Rajiv has created a branch with the booking work
